<?xml version="1.0"?>

<launch>
	<!-- SENSORS -->
	<arg name="sensor_calibration_file" default=""/>
	<!-- only needed if there are multiple ncameras/lidars/imus/odometries in the calibration file. -->
	<arg name="selected_ncamera_sensor_id" default=""/>
	<arg name="selected_imu_sensor_id" default=""/>
	<arg name="selected_lidar_sensor_id" default=""/>
	<arg name="selected_odometry_6dof_sensor_id" default=""/>
	<arg name="imu_to_camera_time_offset_ns" default="0"/>
	<arg name="vio_nframe_sync_tolerance_ns" default="500000"/>
	<arg name="vio_nframe_sync_max_output_frequency_hz" default="10"/>


	<!-- DATA SOURCE -->
	<arg name="datasource_type" default="rosbag"/>
	<arg name="datasource_rosbag" default=""/>

	<!-- DATA PREPROCESSING -->
	<arg name="image_apply_clahe_histogram_equalization" default="false"/>

	<!-- LOCALIZATION BASED ON VISION -->
	<arg name="enable_visual_localization" default="false"/>
	<arg name="visual_localization_map_folder" default=""/>

	<!-- LOCALIZATION BASED ON LIDAR -->
	<arg name="enable_lidar_localization" default="false"/>
	<arg name="lidar_localization_map_folder" default=""/>

	<!-- ONLINE MAPPING -->
	<arg name="enable_online_mapping" default="false"/>

	<!-- MAP BUILDING -->
	<arg name="map_output_folder" default=""/>
	<arg name="map_save_on_shutdown" default="false"/>
	<arg name="map_overwrite_enabled" default="false"/>
	<arg name="map_compute_localization_map" default="false"/>
	<arg name="map_run_keyframing_when_saving" default="false"/>
	<arg name="map_remove_bad_landmarks" default="false"/>
	<arg name="map_initialize_and_triangulate_landmarks_when_saving" default="true"/>
	<arg name="map_save_every_n_sec" default="0"/>
	<arg name="map_split_map_into_submaps_when_saving_periodically" default="false"/>
  <arg name="map_add_odometry_edges_if_less_than_n_common_landmarks" default="70"/>


	<!-- ATTACHING RESOURCES TO MAP -->
	<arg name="map_builder_save_image_as_resources" default="false"/>
	<arg name="map_builder_save_point_clouds_as_resources" default="false"/>
	<arg name="map_builder_save_point_clouds_as_range_image_camera_id" default=""/>
	<arg name="map_builder_save_point_clouds_as_range_image_including_intensity_image" default="true"/>
	<arg name="map_builder_save_point_cloud_maps_as_resources" default="true"/>

	<!-- TF -->
	<arg name="tf_map_frame" default="map"/>
	<arg name="tf_mission_frame" default="mission"/>
	<arg name="tf_imu_frame" default="imu"/>
	<arg name="tf_imu_refined_frame" default="imu_refined"/>

	<!-- DEBUGGING AND LOGGING -->
	<arg name="alsologtostderr" default="true"/>
	<arg name="colorlogtostderr" default="true"/>
	<arg name="v" default="0"/>
	<arg name="publish_debug_markers" default="true"/>
	<arg name="vis_default_namespace" default="maplab_rviz"/>
	<arg name="export_estimated_poses_to_csv" default=""/>

  <arg name="detection_visualize_keypoints" default="false"/>
  <arg name="feature_tracker_visualize_feature_tracks" default="false"/>
  <arg name="feature_tracker_visualize_keypoint_matches" default="false"/>
  <arg name="feature_tracker_visualize_keypoints" default="false"/>
  <arg name="feature_tracker_visualize_keypoints_individual_frames" default="false"/>
  <arg name="vio_camera_topic_suffix" default="/image_raw" />

	<node name="maplab_node" pkg="maplab_node" type="maplab_node" output="screen" clear_params="true">

		<!-- SENSORS -->
		<param name="sensor_calibration_file" value="$(arg sensor_calibration_file)"/>
		<param name="selected_ncamera_sensor_id" value="$(arg selected_ncamera_sensor_id)"/>
		<param name="selected_imu_sensor_id" value="$(arg selected_imu_sensor_id)"/>
		<param name="selected_lidar_sensor_id" value="$(arg selected_lidar_sensor_id)"/>
		<param name="selected_odometry_6dof_sensor_id" value="$(arg selected_odometry_6dof_sensor_id)"/>

		<!-- DATA SOURCE -->
		<param name="datasource_rosbag" value="$(arg datasource_rosbag)"/>
		<param name="datasource_type" value="$(arg datasource_type)"/>
		<param name="imu_to_camera_time_offset_ns" value="$(arg imu_to_camera_time_offset_ns)"/>
		<param name="vio_nframe_sync_tolerance_ns" value="$(arg vio_nframe_sync_tolerance_ns)"/>
		<param name="vio_nframe_sync_max_output_frequency_hz" value="$(arg vio_nframe_sync_max_output_frequency_hz)"/>

		<!-- DATA PREPROCESSING -->
		<param name="image_apply_clahe_histogram_equalization" value="$(arg image_apply_clahe_histogram_equalization)"/>

		<!-- VISION -->
		<param name="enable_visual_localization" value="$(arg enable_visual_localization)"/>
		<param name="visual_localization_map_folder" value="$(arg visual_localization_map_folder)"/>

		<!-- LIDAR -->
		<param name="enable_lidar_localization" value="$(arg enable_lidar_localization)"/>
		<param name="lidar_localization_map_folder" value="$(arg lidar_localization_map_folder)"/>

		<!-- ONLINE MAPPING -->
		<param name="enable_online_mapping" value="$(arg enable_online_mapping)"/>

		<!-- MAP -->
		<param name="map_output_folder" value="$(arg map_output_folder)"/>
		<param name="map_save_on_shutdown" value="$(arg map_save_on_shutdown)"/>
		<param name="map_overwrite_enabled" value="$(arg map_overwrite_enabled)"/>
		<param name="map_compute_localization_map" value="$(arg map_compute_localization_map)"/>
		<param name="map_run_keyframing_when_saving" value="$(arg map_run_keyframing_when_saving)"/>
		<param name="vi_map_landmark_quality_min_observers" value="2" if="$(eval map_remove_bad_landmarks)"/>
		<param name="map_remove_bad_landmarks" value="$(arg map_remove_bad_landmarks)"/>
		<param name="map_initialize_and_triangulate_landmarks_when_saving" value="$(arg map_initialize_and_triangulate_landmarks_when_saving)"/>
		<param name="map_save_every_n_sec" value="$(arg map_save_every_n_sec)"/>
		<param name="map_split_map_into_submaps_when_saving_periodically" value="$(arg map_split_map_into_submaps_when_saving_periodically)"/>
	  <param name="map_add_odometry_edges_if_less_than_n_common_landmarks" value="$(arg map_add_odometry_edges_if_less_than_n_common_landmarks)"/>

		<!-- ATTACHING RESOURCES TO MAP -->
		<param name="map_builder_save_image_as_resources" value="$(arg map_builder_save_image_as_resources)"/>
		<param name="map_builder_save_point_clouds_as_resources" value="$(arg map_builder_save_point_clouds_as_resources)"/>
		<param name="map_builder_save_point_clouds_as_range_image_camera_id" value="$(arg map_builder_save_point_clouds_as_range_image_camera_id)"/>
		<param name="map_builder_save_point_clouds_as_range_image_including_intensity_image" value="$(arg map_builder_save_point_clouds_as_range_image_including_intensity_image)"/>
		<param name="map_builder_save_point_cloud_maps_as_resources" value="$(arg map_builder_save_point_cloud_maps_as_resources)"/>

		<!-- TF -->
		<param name="tf_map_frame" value="$(arg tf_map_frame)"/>
		<param name="tf_mission_frame" value="$(arg tf_mission_frame)"/>
		<param name="tf_imu_frame" value="$(arg tf_imu_frame)"/>
		<param name="tf_imu_refined_frame" value="$(arg tf_imu_refined_frame)"/>

		<!-- DEBUGGING AND LOGGING -->
		<param name="alsologtostderr" value="$(arg alsologtostderr)"/>
		<param name="colorlogtostderr" value="$(arg colorlogtostderr)"/>
		<param name="v" value="$(arg v)"/>
		<param name="publish_debug_markers" value="$(arg publish_debug_markers)"/>
		<param name="vis_default_namespace" value="$(arg vis_default_namespace)"/>

		<param name="export_estimated_poses_to_csv" value="$(arg export_estimated_poses_to_csv)"/>
		<param name="detection_visualize_keypoints" value="$(arg detection_visualize_keypoints)"/>
		<param name="feature_tracker_visualize_feature_tracks" value="$(arg feature_tracker_visualize_feature_tracks)"/>
		<param name="feature_tracker_visualize_keypoint_matches" value="$(arg feature_tracker_visualize_keypoint_matches)"/>
		<param name="feature_tracker_visualize_keypoints" value="$(arg feature_tracker_visualize_keypoints)"/>
		<param name="feature_tracker_visualize_keypoints_individual_frames" value="$(arg feature_tracker_visualize_keypoints_individual_frames)"/>
		<param name="vio_camera_topic_suffix" value="$(arg vio_camera_topic_suffix)"/>
	</node>
</launch>
